File: src/components/Notion.js
import express from "express";
import axios from "axios";
import cors from "cors";
import dotenv from "dotenv";

dotenv.config();

const app = express();
app.use(cors());
app.use(express.json());

// Load Notion API settings from .env
const NOTION_API_KEY = process.env.NOTION_API_KEY
const DATABASE_ID = process.env.DATABASE_ID

const NOTION_HEADERS = {
  Authorization: `Bearer ${NOTION_API_KEY}`,
  "Content-Type": "application/json",
  "Notion-Version": "2022-06-28",
};

// --- Add Bookmark ---
app.post("/add_bookmark", async (req, res) => {
  const { verse, content } = req.body;

  if (!verse || !content) {
    return res.status(400).json({ error: "Both 'verse' and 'content' are required" });
  }

  const payload = {
    parent: { database_id: BOOKMARKS_DB_ID },
    properties: {
      Verse: {
        title: [{ text: { content: verse } }],
      },
      Content: {
        rich_text: [{ text: { content: content } }],
      },
    },
  };

  try {
    const response = await axios.post("https://api.notion.com/v1/pages", payload, {
      headers: NOTION_HEADERS,
    });
    res.json({ success: true, notion_response: response.data });
  } catch (err) {
    res.status(err.response?.status || 500).json({
      success: false,
      error: err.response?.data || err.message,
    });
  }
});

// --- Get All Bookmarks ---
app.get("/bookmarks", async (req, res) => {
  try {
    const response = await axios.post(
      `https://api.notion.com/v1/databases/${BOOKMARKS_DB_ID}/query`,
      {},
      { headers: NOTION_HEADERS }
    );

    const results = response.data.results.map((page) => {
      const verse =
        page.properties?.Verse?.title?.[0]?.text?.content || "";
      const content =
        page.properties?.Content?.rich_text?.[0]?.text?.content || "";
      return { verse, content };
    });

    res.json(results);
  } catch (err) {
    res.status(err.response?.status || 500).json({
      error: err.response?.data || err.message,
    });
  }
});

// --- Start server ---
const PORT = process.env.PORT || 5000;
app.listen(PORT, () => {
  console.log(`✅ Server running on http://localhost:${PORT}`);
});

File: src/components/Notion.py
# Runs in https://www.pythonanywhere.com/
# Migrate to DigitalOcean
from flask import Flask, request, jsonify
import requests
import os
from flask_cors import CORS

from dotenv import load_dotenv

# Load the .env file
load_dotenv()

app = Flask(__name__)

# Allow requests from your frontend domain (e.g. localhost or production)
CORS(app, resources={r"/*": {"origins": "*"}})

NOTION_API_KEY = os.getenv("NOTION_API_KEY")
DATABASE_ID = os.getenv("DATABASE_ID")

NOTION_HEADERS = {
    "Authorization": f"Bearer {NOTION_API_KEY}",
    "Notion-Version": "2022-06-28",
    "Content-Type": "application/json",
}


def verse_exists(verse_name):
    """Check if the verse already exists in the Notion database."""
    query_url = f"https://api.notion.com/v1/databases/{DATABASE_ID}/query"
    query_payload = {
        "filter": {
            "property": "Verse",
            "rich_text": {
                "equals": verse_name
            }
        }
    }
    response = requests.post(query_url, headers=NOTION_HEADERS, json=query_payload)
    if response.status_code == 200:
        results = response.json().get("results", [])
        return len(results) > 0
    else:
        print("Error querying Notion:", response.text)
        return False


@app.route("/check-verse", methods=["POST"])
def check_verse():
    data = request.json
    verse = data.get("verse")

    if not verse:
        return jsonify({"error": "Missing 'verse' parameter"}), 400

    exists = verse_exists(verse)
    return jsonify({"exists": exists})


@app.route("/add-verse", methods=["POST"])
def add_verse():
    data = request.json
    verse = data.get("verse")
    content = data.get("content")

    if not verse or not content:
        return jsonify({"error": "Both 'verse' and 'content' fields are required"}), 400

    if verse_exists(verse):
        return jsonify({"message": f"'{verse}' already exists, skipping."}), 200

    payload = {
        "parent": {"database_id": DATABASE_ID},
        "properties": {
            "Verse": {"title": [{"text": {"content": verse}}]},
            "Content": {"rich_text": [{"text": {"content": content}}]},
        },
    }

    response = requests.post("https://api.notion.com/v1/pages", headers=NOTION_HEADERS, json=payload)

    if response.status_code == 200:
        return jsonify({"success": True, "message": f"'{verse}' added successfully!"}), 200
    else:
        return jsonify({"error": response.text}), response.status_code


@app.route("/")
def home():
    return "Notion Verse API is running!"
